src/mscorlib/corefx/System/IO/PathHelper.Windows.cs

```diff
@@ -271,119 +268,124 @@ private static string TryExpandShortFileName(StringBuffer outputBuffer, string o
+                        // Look to see if we couldn't find the file
+                        int error = Marshal.GetLastWin32Error();
+                        if (error != Interop.mincore.Errors.ERROR_FILE_NOT_FOUND && error != Interop.mincore.Errors.ERROR_PATH_NOT_FOUND)
+                        {
+                            // Some other failure, give up
+                            break;
+                        }
 
-                    // We couldn't find the path at the given index, start looking further back in the string.
-                    foundIndex--;
+                        // We couldn't find the path at the given index, start looking further back in the string.
+                        foundIndex--;
 
-                    for (; foundIndex > rootLength && inputBuffer[foundIndex] != '\\'; foundIndex--) ;
-                    if (foundIndex == rootLength)
+                        for (; foundIndex > rootLength && inputBuffer[foundIndex] != '\\'; foundIndex--) ;
+                        if (foundIndex == rootLength)
+                        {
+                            // Can't trim the path back any further
+                            break;
+                        }
+                        else
+                        {
+                            // Temporarily set a null in the string to get Windows to look further up the path
+                            inputBuffer[foundIndex] = '\0';
+                        }
+                    }
+                    else if (result > outputBuffer.Capacity)
                     {
-                        // Can't trim the path back any further
-                        break;
+                        // Not enough space. The result count for this API does not include the null terminator.
+                        outputBuffer.EnsureCapacity(checked((int)result));
+                        result = Interop.mincore.GetLongPathNameW(inputBuffer.UnderlyingArray, outputBuffer.UnderlyingArray, (uint)outputBuffer.Capacity);
                     }
                     else
                     {
-                        // Temporarily set a null in the string to get Windows to look further up the path
-                        inputBuffer[foundIndex] = '\0';
+                        // Found the path
+                        success = true;
+                        outputBuffer.Length = checked((int)result);
+                        if (foundIndex < inputLength - 1)
+                        {
+                            // It was a partial find, put the non-existent part of the path back
+                            outputBuffer.Append(ref inputBuffer, foundIndex, inputBuffer.Length - foundIndex);
+                        }
                     }
                 }
-                else if (result > outputBuffer.CharCapacity)
+
+                // Strip out the prefix and return the string
+                StringBuffer bufferToUse = success ? outputBuffer : inputBuffer;
```

Comment:
Given the current code that follows this, this looks safe, but it's also quite fragile, as now any mutation to bufferToUse that causes it to expand could result in freeing a buffer back to the array pool twice, once when bufferToUse expands, and then once when inputBuffer is freed in the finally.